crates/core/src/delta_datafusion/find_files/mod.rs

```diff
@@ -70,81 +85,199 @@ impl QueryPlanner for FindFilesPlanner {
+        .select(vec![col(PATH_COLUMN)])?;
+    let df_schema = df.schema().clone();
+    let batches = df.collect().await?;
+    Ok(concat_batches(&SchemaRef::from(df_schema), &batches)?)
+}
+
+async fn scan_table_by_files(
+    snapshot: DeltaTableState,
+    log_store: LogStoreRef,
+    state: SessionState,
+    expression: Expr,
+) -> Result<RecordBatch> {
+    register_store(log_store.clone(), state.runtime_env().clone());
+    let scan_config = DeltaScanConfigBuilder {
+        include_file_column: true,
+        ..Default::default()
+    }
+    .build(&snapshot)?;
+
+    let logical_schema = df_logical_schema(&snapshot, &scan_config)?;
+
+    // Identify which columns we need to project
+    let mut used_columns = expression
         .to_columns()?
         .into_iter()
-        .map(Expr::Column)
-        .collect::<Vec<_>>();
-
-    if let Some(column) = batch.column_by_name(PATH_COLUMN) {
-        let mut column_iter = column.as_string::<i32>().into_iter();
-        while let Some(Some(row)) = column_iter.next() {
-            let df = ctx
-                .read_parquet(row, ParquetReadOptions::default())
-                .await?
-                .select(columns.clone())?
-                .filter(predicate.clone())?
-                .limit(0, Some(1))?;
-            if df.count().await? > 0 {
-                batches.push(row);
-            }
-        }
-    }
-    let str_array = Arc::new(StringArray::from(batches));
-    RecordBatch::try_new(ONLY_FILES_SCHEMA.clone(), vec![str_array]).map_err(Into::into)
+        .map(|column| logical_schema.index_of(&column.name))
+        .collect::<std::result::Result<Vec<usize>, ArrowError>>()?;
+    // Add path column
+    used_columns.push(logical_schema.index_of(scan_config.file_column_name.as_ref().unwrap())?);
+
+    let scan = DeltaScanBuilder::new(&snapshot, log_store, &state)
+        .with_filter(Some(expression.clone()))
+        .with_projection(Some(&used_columns))
+        .with_scan_config(scan_config)
+        .build()
+        .await?;
+
+    let scan = Arc::new(scan);
+    let input_schema = scan.logical_schema.as_ref().to_owned();
+    let input_dfschema = input_schema.clone().try_into()?;
+
+    let predicate_expr = create_physical_expr(
+        &Expr::IsTrue(Box::new(expression.clone())),
+        &input_dfschema,
+        state.execution_props(),
+    )?;
+
+    let filter: Arc<dyn ExecutionPlan> =
+        Arc::new(FilterExec::try_new(predicate_expr, scan.clone())?);
+    let limit: Arc<dyn ExecutionPlan> = Arc::new(LocalLimitExec::new(filter, 1));
+    let field_idx = input_schema.index_of(PATH_COLUMN)?;
+    let task_ctx = Arc::new(TaskContext::from(&state));
+    let path_batches: Vec<RecordBatch> = datafusion::physical_plan::collect(limit, task_ctx)
+        .await?
+        .into_iter()
+        .map(|batch| batch.project(&[field_idx]).unwrap())
+        .collect();
+
+    let result_batches = concat_batches(&ONLY_FILES_SCHEMA.clone(), &path_batches)?;
```

Comment:
I was thinking for this operation to output file paths as soon as they are discovered to allow operations downstream to begin their work ASAP. When performing an memory only scan to makes sense to output as a single batch since IO is minimal.
For this PR I think it's okay since it aligns with the previous behaviour but its something we can change.

Reply:
can you elaborate a bit more? I understand what you are suggesting here, but I'm not sure what you are expecting. Like you want the return type to be a vec or record batches or a stream or something?

Reply:
Yes output record batches and these record batches should simply be a string. There is no benefit for using dictionaries in the output.

Reply:
Oh you're talking about the actual schema of the record batch, I thought you were saying the return type of the function was wrong or something of the sort for it to be "immediately available" or something of the sort.